rust/src/dataset/optimize.rs

```diff
@@ -0,0 +1,779 @@
+        .unwrap_or(0);
+
+    let mut new_fragments: HashMap<Range<usize>, Vec<Fragment>> = HashMap::new();
+
+    while let Some(result) = result_stream.next().await {
+        let mut result = result?;
+
+        metrics += result.metrics;
+
+        // Assign fragment ids
+        for fragment in &mut result.new_fragments {
+            fragment.id = current_fragment_id;
+            current_fragment_id += 1;
+        }
+
+        let RewriteResult {
+            new_fragments: fragments,
+            replace_range,
+            ..
+        } = result;
+        new_fragments.insert(replace_range, fragments);
+    }
+
+    let final_fragments = compaction_plan.build_fragment_list(new_fragments);
+
+    // Commit the dataset transaction
+    let indices = if metrics.fragments_removed == dataset.get_fragments().len() {
+        // All fragments were replaced, so the index is totally invalid
+        None
+    } else {
+        Some(dataset.load_indices().await?)
+    };
+
+    // TODO: replace this with from_previous
+    let mut manifest = Manifest::new(dataset.schema(), Arc::new(final_fragments));
+
+    manifest.version = dataset
+        .latest_manifest()
+        .await
+        .map(|m| m.version + 1)
+        .unwrap_or(1);
+
+    write_manifest_file(
+        &dataset.object_store,
+        &dataset.base,
+        &mut manifest,
+        indices,
+        Default::default(),
+    )
+    .await?;
+
+    dataset.manifest = Arc::new(manifest);
+
+    Ok(metrics)
+}
+
+// TODO: ideally these metrics should already be in the manifest, so we don't
+// have to scan during compaction.
+
+/// Information about a fragment used to decide it's fate in compaction
+#[derive(Debug)]
+struct FragmentMetrics {
+    /// The number of original rows in the fragment
+    pub fragment_length: usize,
+    /// The number of rows that have been deleted
+    pub num_deletions: usize,
+}
+
+impl FragmentMetrics {
+    /// The fraction of rows that have been deleted
+    fn deletion_percentage(&self) -> f32 {
+        if self.fragment_length > 0 {
+            self.num_deletions as f32 / self.fragment_length as f32
+        } else {
+            0.0
+        }
+    }
+}
+
+async fn collect_metrics(fragment: &FileFragment) -> Result<FragmentMetrics> {
+    let fragment_length = fragment.fragment_length();
+    let num_deletions = fragment.count_deletions();
+    let (fragment_length, num_deletions) =
+        futures::future::try_join(fragment_length, num_deletions).await?;
+    Ok(FragmentMetrics {
+        fragment_length,
+        num_deletions,
+    })
+}
+
+/// A plan for what groups of fragments to compact. To preserve order, we only
+/// compact groups of fragments that are adjacent in the dataset. So each group
+/// to compact is represented as a range of indices in the `fragments` list. We
+/// also track the ranges of fragments that should be kept as-is, to make it
+/// easier to build the final list of fragments.
+struct CompactionPlan {
```

Comment:
If we consider "rewrite for better encoding", this is not just a compact plan, right? It is like a `Optimize(Storage)Plan`.